### ISSUE TITLE
Изменение высоты тона при изменении скорости воспроизведения

### ISSUE THREAD

**Issue by IAmKapuze** [anilibria/anilibria-winmaclinux#48]
Если в плеере изменить скорость воспроизведения, то меняется и высота тона, что не очень хорошо сказывается на восприятии.
Я не сильно глубоко смотрел в код, но, как я понял, для отображения плеера используется html со встроенным clappr и плагин clappr-playback-rate-plugin к нему. Насколько я понимаю, все эти плагины работают путем изменения свойства HTMLMediaElement.playbackRate. На [странице об HTMLMediaElement.playbackRate с MDN](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement/playbackRate) по этому поводу написано:

> The pitch of the audio is corrected by default and is the same for every speed. Some browsers implement the non-standard HTMLMediaElement.preservesPitch property to control this.

Также на [странице об HTMLMediaElement с MDN](https://developer.mozilla.org/en-US/docs/Web/API/HTMLMediaElement) написано:

> Is a Boolean that determines if the pitch of the sound will be preserved. If set to false, the pitch will adjust to the speed of the audio. This is implemented with prefixes in Firefox (mozPreservesPitch) and WebKit (webkitPreservesPitch).

Если всё так, как я предполагаю (что не факт), то либо preservesPitch находится в значении false, либо это регулируется как-то иначе. Исправимо ли это?

**Comment by trueromanus** [comment]
Здравствуйте, тут надо было уточнить что Вы о экспериментальном простеньком web клиенте для приложения, потому что само приложение не имеет никакого отношения к вебу в принципе. 
> Если в плеере изменить скорость воспроизведения, то меняется и высота тона, что не очень хорошо сказывается на восприятии.  

Если честно не понял о чем речь. Я сейчас проверил и скорость аудио меняется нормально.  
Проверил **HTMLMediaElement.preservesPitch**  и оно равно true, в этом легко убедиться открыв консоль.
Я проверял на Chrome Windows 10.  
  
А вообще я не специалист по web плееру, поэтому взял уже готовую библиотеку.    
И тут еще стоит понимать тот факт что я сделал пример как можно создать клиент, т.е. я не предполагал что именно он будет использоваться как финальный вариант. Поэтому стоит воспринимать это как учебный пример как можно сделать клиент для приложения. И если есть желание поразвивать его или замутить свой то welcome :).
Сам файл находиться здесь https://github.com/anilibria/anilibria-win/blob/master/videoplayer.html и так кода на 3 копейки.

**Comment by IAmKapuze** [comment]
Так, а что есть "экспериментальный простенький web-клиент"? Я говорю именно [о клиенте winmaclinux](https://github.com/anilibria/anilibria-winmaclinux), так как сам использую linux. В плеере у этого клиента есть проблема с питчем. На Windows в том числе. Однако, сейчас проверил [именно win клиент](https://github.com/anilibria/anilibria-win), там такой проблемы действительно нет.
Возможно, я неправильно понял суть работы приложения, и меня спутали файлы из директории webplayer. Рассуждал я как:
В плеере есть кнопка, например, "x0.25". Ищу её по лейблу, вижу её в OnlinePlayer.qml. Вижу, что она вызывает функцию setVideoSpeed(0.25)
В setVideoSpeed происходит вызов remotePlayer.broadcastCommand с соответствующей командой, которая отправляется куда-то на веб-сокет.
videoplayer.html как раз коннектится к какому-то сокету и обрабатывает команду вида "playbackratechanged", которая в OnlinePlayer.qml и упаковывается.
Но при этом изменения в файле videoplayer.html никакого результата не вносят, так что я предположил, что они просто дублируют страницу, которой отвечает сервер (для понимания сути происходящего), а сам код страницы менять нужно на стороне сервера.
Если подскажите как на самом деле работает приложение (в плане изменения скорости воспроизведения) и куда копать чтобы зафиксировать питч при изменении скорости - буду сильно благодарен. Коммит соответствующий также внесу :)

**Comment by trueromanus** [comment]
"В плеере у этого клиента есть проблема с питчем." - дак что за проблема то? Может видео записать?

Установка скорости происходит на [этой строке](https://github.com/anilibria/anilibria-winmaclinux/blob/4ebe42cb54567c03455d7964efb645470eec477b/src/Views/OnlinePlayer.qml#L1453)
Остальная магия происходит уже внутри контрола MediaPlayer. Но насколько я знаю он просто использует низлежайший интерфейс из системы.

Какой конкретно linux? В linux используется gstreamer установленный в системе, может быть проблема в его настройке.  
Дока по этому параметру тут:
https://doc.qt.io/qt-5/qml-qtmultimedia-mediaplayer.html#playbackRate-prop

**Comment by IAmKapuze** [comment]
https://user-images.githubusercontent.com/16070551/120922811-ece8a200-c6d3-11eb-85a4-f83702f342c4.mp4

https://user-images.githubusercontent.com/16070551/120922949-bfe8bf00-c6d4-11eb-88bc-e2abea4246aa.mp4

О том какой конкретно linux:
Linux OS: Manjaro 21.0.5 Ornara
Kernel: x86_64 Linux 5.10.36-2-MANJARO
DE: KDE 5.82.0 / Plasma 5.21.5
gstreamer: 1.18.4-1

Однако, не думаю, что это всё имеет значение, так как, как я и написал выше, на Windows проблема тоже присутствует. Где смотрел не туда - понял. Посмотрю что можно сделать. Спасибо

**Comment by trueromanus** [comment]
Посмотрел видео и не понял в чем проблема, видимо потому что я никогда не пользуюсь скоростями отличными от обычной.

**Comment by IAmKapuze** [comment]
Проблема в том, что появляется эффект питча. Записал пример нормального поведения [с клиента anilibria-win](https://github.com/anilibria/anilibria-win):

https://user-images.githubusercontent.com/16070551/120995952-912f1f00-c78e-11eb-96ff-0eda03bc4b7f.mp4

**Comment by buggyman** [comment]
> Если в плеере изменить скорость воспроизведения, то меняется и высота тона, что не очень хорошо сказывается на восприятии.

Во всех плеерах такое происходит, такова природа физики звука. Или запишите видео того же куска в другом плеере, чтобы наглядно показать чего вы хотите.

**Comment by IAmKapuze** [comment]
> > Если в плеере изменить скорость воспроизведения, то меняется и высота тона, что не очень хорошо сказывается на восприятии.
> 
> Во всех плеерах такое происходит, такова природа физики звука. Или запишите видео того же куска в другом плеере, чтобы наглядно показать чего вы хотите.

1. Перед тем как просить меня записать видео в другом плеере, сначала посмотрите хотя бы последний пост в теме, который находится прямо над вашим.
2. Перед тем как писать мне о природе звука, почитайте о таком понятии как pitch compensation или тоновая компенсация.
Вы в VLC никогда скорость воспроизведения не увеличивали? А в Youtube пробовали ставить значение отличное от 1x ? (И ещё раз напомню про свой пост прямо над вашим, где в клиенте для win10 прекрасно работает компенсация высоты тона)
И вообще, раз уж тема зашла, природа звука здесь не при чем. Звук в компьютере находится в цифровом виде. Колонки гелием никто не накачивает при изменении скорости :)

**Comment by buggyman** [comment]
> 1. Перед тем как просить меня записать видео в другом плеере, сначала посмотрите хотя бы последний пост в теме, который находится прямо над вашим.

Пардон, в первый раз почему-то не запускалось это видео, поэтому решил что это картинка.

> 2. Перед тем как писать мне о природе звука, почитайте о таком понятии как pitch compensation или тоновая компенсация.

Ок.

**Comment by niallipa** [comment]
На самом деле столкнулся пару раз с тем, что голос запитчен был без изменение скорости, что это вызвало - без понятия, но было смешно смотреть аниме с писклявыми голосами, помогает только перезапуск, скорости не меняю никогда и сказать что вызывает такое тоже не могу, баг редкий

**Comment by trueromanus** [comment]
> На самом деле столкнулся пару раз с тем, что голос запитчен был без изменение скорости, что это вызвало - без понятия, но было смешно смотреть аниме с писклявыми голосами, помогает только перезапуск, скорости не меняю никогда и сказать что вызывает такое тоже не могу, баг редкий
> 
> Можете воспринимать как оффтоп

Проблема с питчем из-за того что в плеере приложения не применяется аудио фильтров для компенсации питча и я не настолько крут чтобы написать их с нуля, я пробовал готовые опенсорсные решения но они не решали проблему целиком. В версии AniLibria.Qt для qt6 интегрирован другой плеер в котором эта проблема решена на уровне плеера. Пока версия для qt6 в процессе разработки :(

**Comment by trueromanus** [comment]
Будет реализовано в #74

**Comment by trueromanus** [comment]
После перехода на VLC в версии 1.2.4 проблема решиться.
